### ISSUE TITLE
Keep hyperlinks in the original document

### ISSUE THREAD

**Issue by CsatiZoltan** [xournalpp/xournalpp#1873]
I am not sure if is related to #1027 . Pdf of scientific papers contain a lot of *www* and internal references. Once I edit the pdf (adding LaTeX notes) and export it as a pdf, the original links are lost.

**Comment by Technius** [comment]
This is something that I would like to implement, but it will take some work. In an earlier PR, I added the ability to preserve document outlines in the PDF (#1741), which has been made possible to due new APIs in Cairo 1.16. Unfortunately, Poppler's glib frontend only has the ability to read PDFs (but not render/write), so we're stuck with the problem that we basically can only preserve things that we are aware of and have written code for.

**Comment by idotobi** [comment]
@Technius : Is this the relevant Issue for poppler? https://gitlab.freedesktop.org/poppler/poppler/-/issues/677

**Comment by Technius** [comment]
We have to manually ensure that original links are exported. This can be done with `poppler-glib` functions such as the ones in #2660.

**Comment by bgrenet** [comment]
As a (temporary) workaround, I am quite satisfied using `qpdf` for that task. My (automated) process is as follows:

1. Export the .xopp file to PDF without the background: `xournalpp <INPUT> -p <TEMP> --export-no-background`
2. Overlay the <TEMP> export with the original background, using qpdf: `qpdf <BACKGROUND> --overlay <TEMP> -- <OUTPUT>`
3. Remove the <TEMP> file: `rm <TEMP>`.

The resulting <OUTPUT> pdf file has all the hyperlinks, bookmarks, etc.

**Comment by tattsan** [comment]
It is a nice workaround, @bgrenet !
However, this method does not allow one to insert an additional xopp page between background-PDF pages. `qpdf`'s command line could be a bit more complex to handle this case, but it would be difficult to write an automated script.

**Comment by bgrenet** [comment]
Indeed, I faced a similar issue: I am annotating slides and I need sometimes to duplicate some pages (rather than inserting new xopp pages). Using qpdf, it is quite easy to perform the duplication on the <BACKGROUND> if we know the pages to duplicate. Now, the issue is to automatize the process (either add an argument for duplication to my script or better directly detect the duplication). But I am reluctant to work too hard on this since the goal is a temporary workaround.

**Comment by rolandlo** [comment]
It should not be too hard to write a Lua plugin (using `qpdf` and `pdftk`) for the hyperlink-preserving export. Consider the following steps:
1) export document to pdf without background (only pages with pdf background) -> `export1.pdf`
2) export document to pdf with background (only pages without pdf background) -> `export2.pdf`
3) Use `qpdf` to overlay pages from the background pdf with the correct pages from `export1.pdf` -> `overlayed.pdf`
4) Use `pdftk` to combine `overlayed.pdf` with the missing pages from `export2.pdf` while removing pages from `overlayed.pdf` that are not used in the document

Say the document has 10 pages, where 1,2,3,4,7,8,10 have PDF background (`bg.pdf`) corresponding to pages 1-7 in the PDF background and 5,6,9 don't have PDF background. No PDF pages are missing in the document (for simplicity). Then something like
```lua
app.export({["outputFile"] = "export1.pdf", ["range"] = "1,2,3,4,7,8,10", ["background"] = "none"})
app.export({["outputFile"] = "export2.pdf", ["range"] = "5,6,9")
os.execute("qpdf bg.pdf --overlay export1.pdf -- overlayed.pdf")
os.execute("pdftk A=overlayed.pdf B=export2.pdf cat A1 A2 A3 A4 B5 B6 A7 A8 B9 A10 output final.pdf")
```
would do the job.

**Comment by bkuczenski** [comment]
It is beyond bonkers that I cannot open a PDF, make a change, and save it without destroying hyperlinks! How is this 2023??

**Comment by rolandlo** [comment]
@bkuczenski PRs are welcome.

**Comment by bgrenet** [comment]
As a follow up to my previous comment, I wrote a pretty basic python script to automate the export. It requires a few standard Python3 libraries, as well as [QPDF](https://qpdf.sourceforge.io/) (must be available from the command line `qpdf`. Its main characteristics:

- supports any removed or duplicated page;
- only supports (for now) one single pdf background;
- preserves the metadata of the PDF, the hyperlinks, the bookmarks, etc.

Feel free to propose some improvements!

**Comment by CsatiZoltan** [comment]
@bhennion Will #6364 be part of an upcoming release, or Xournal++ will have to be built from the source for this feature?

**Comment by bhennion** [comment]
It will be part of v1.3.0. The upcoming release will be 1.2.8 and will only contain bugfixes. I'm not quite sure of when we'll release 1.3.0.

In any case, you can already use our nigthly builds, which do contain the feature already: https://github.com/xournalpp/xournalpp/releases/tag/nightly

**Comment by CsatiZoltan** [comment]
I downloaded the [latest version](https://github.com/xournalpp/xournalpp/releases/download/nightly/xournalpp-1.2.7+dev-nightly-20250714-x86_64.AppImage). I opened a PDF, drew a few lines, exported it as PDF, and the links are lost.

**Comment by bhennion** [comment]
> I downloaded the [latest version](https://github.com/xournalpp/xournalpp/releases/download/nightly/xournalpp-1.2.7+dev-nightly-20250714-x86_64.AppImage). I opened a PDF, drew a few lines, exported it as PDF, and the links are lost.

Thanks for signaling that! Indeed, I forgot to enable the feature in the releases... It should be fixed. The next nightly build (tonight) should fix it.

**Comment by CsatiZoltan** [comment]
I confirm that it works now. Thank you a lot for implementing this feature! I can now annotate journal articles, which previously I couldn't due to the loss of the hyperlinks.

**Comment by ngirard** [comment]
> I can now annotate journal articles, which previously I couldn't due to the loss of the hyperlinks.

Hey @CsatiZoltan ,

just curious: why preferring Xournal++ over Zotero for this use case ?

**Comment by CsatiZoltan** [comment]
@ngirard I don't think I can add LaTeX notes and handwriting on PDFs in Zotero, both of them being important for me. Moreover, Zotero does many things, while I prefer having separate, specialized tools.
